### PR TITLE
Feature/contract request handler

### PR DESCRIPTION
**Pull request by volodymyr-basiuk**

### CODE REVIEW

File: src/iden3comm/types/protocol/contract-request.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import { Signer } from 'ethers';
+import { ZeroKnowledgeProofRequest } from './auth';
+
+/** ContractInvokeRequest represents structure of contract invoke request object */
+export type ContractInvokeRequest = {
```
[the remark this answers is not in the document]
Comment (volodymyr-basiuk) [reply src/iden3comm/types/protocol/contract-request.ts:5]: inside protocol models we use `type`

File: src/storage/blockchain/onchain-zkp-verifier.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    for (const zkProof of zkProofResponses) {
+      const requestID = zkProof.id;
+      const inputs = zkProof.pub_signals;
+
+      const payload = [
+        requestID,
+        inputs,
+        zkProof.proof.pi_a.slice(0, 2),
+        [
+          [zkProof.proof.pi_b[0][1].toString(), zkProof.proof.pi_b[0][0].toString()],
```
[the remark this answers is not in the document]
Comment (volodymyr-basiuk) [reply src/storage/blockchain/onchain-zkp-verifier.ts]: no, removed

File: src/iden3comm/handlers/contract-request.ts [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    did: DID,
+    request: Uint8Array,
+    opts?: ContractInvokeHandlerOptions
+  ): Promise<Map<string, ZeroKnowledgeProofResponse>>;
+}
+/**
+ *
+ * Allows to process ContractInvokeRequest protocol message
+ *
+ * @public
```
Comment (vmidyllic) [inline src/iden3comm/handlers/contract-request.ts:48]:
```suggestion
 * @beta
```

File: src/iden3comm/handlers/contract-request.ts [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+  constructor(
+    private readonly _packerMgr: IPackageManager,
+    private readonly _proofService: IProofService,
+    private readonly _zkpVerifier: IOnChainZKPVerifier
+  ) {}
+
+  /**
+   * unpacks contract-request request
+   * @public
```
Comment (vmidyllic) [inline src/iden3comm/handlers/contract-request.ts:75]:
```suggestion
   * @beta
```

File: src/iden3comm/handlers/contract-request.ts [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+   */
+
+  constructor(
+    private readonly _packerMgr: IPackageManager,
+    private readonly _proofService: IProofService,
+    private readonly _zkpVerifier: IOnChainZKPVerifier
+  ) {}
+
+  /**
+   * unpacks contract-request request
```
Comment (vmidyllic) [inline src/iden3comm/handlers/contract-request.ts:74]:
```suggestion
   * unpacks contract-invoke request
```

File: src/iden3comm/handlers/contract-request.ts [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    const ciRequest = message as unknown as ContractInvokeRequest;
+    if (message.type !== PROTOCOL_MESSAGE_TYPE.CONTRACT_INVOKE_REQUEST_MESSAGE_TYPE) {
+      throw new Error('Invalid media type');
+    }
+    return ciRequest;
+  }
+
+  /**
+   * handle contract invoker request
+   * @public
```
Comment (vmidyllic) [inline src/iden3comm/handlers/contract-request.ts:90]:
```suggestion
   * handle contract invoke request
   * @beta
```

File: src/iden3comm/handlers/contract-request.ts [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+    if (!opts.ethSigner) {
+      throw new Error("Can't sign transaction. Provide Signer in options.");
+    }
+
+    const zkRequests = [];
+    for (const proofReq of ciRequest.body.scope) {
+      if (!this._allowedCircuits.includes(proofReq.circuitId as CircuitId)) {
+        throw new Error(
+          `Can't handle circuit ${proofReq.circuitId}. Only onchain circuits allowed.`
```
Comment (vmidyllic) [inline src/iden3comm/handlers/contract-request.ts:115]:
```suggestion
          `Can't handle circuit ${proofReq.circuitId}. Only onchain circuits  are allowed.`
```

File: src/iden3comm/types/protocol/contract-request.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+/** ContractInvokeTransactionData represents structure of contract invoke transaction data object */
+export type ContractInvokeTransactionData = {
+  contract_address: string;
+  method_id: string;
+  chain_id: number;
+  network?: string;
+};
+
+/** ContractInvokeHandlerOptions represents contract invoke handler options */
+export type ContractInvokeHandlerOptions = {
```
Comment (vmidyllic) [inline src/iden3comm/types/protocol/contract-request.ts:29]: I guess we can move it to handler, not a part of protocol type

File: src/storage/interfaces/onchain-zkp-verifier.ts [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+/**
+ * Interface that defines methods for ZKP verifier
+ *
+ * @beta
+ * @interface IOnChainZKPVerifier
+ */
+export interface IOnChainZKPVerifier {
+  /**
+   * Submit ZKP Responses to OnChainZKPVerifier contract.
+   * @public
```
Comment (vmidyllic) [inline src/storage/interfaces/onchain-zkp-verifier.ts:13]:
```suggestion
   * @beta
```

File: src/storage/blockchain/onchain-zkp-verifier.ts [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+ * and submitZKPResponse.
+ *
+ * @beta
+ * @class OnChainZKPVerifier
+ */
+export class OnChainZKPVerifier implements IOnChainZKPVerifier {
+  /**
+   *
+   * Creates an instance of OnChainZKPVerifier.
+   * @public
```
Comment (vmidyllic) [inline src/storage/blockchain/onchain-zkp-verifier.ts:18]:
```suggestion
   * @beta
```

**Review by vmidyllic — APPROVED** [review]

**Pull request merged**